content/docs/digging_deeper/locks.md

```diff
@@ -0,0 +1,357 @@
+
+Following is the list of options the database store accepts (alongside the shared options).
+
+```ts
+{
+  database: stores.database({
+    connectionName: 'postgres',
+    tableName: 'my_locks',
+  }),
+}
+```
+
+<dl>
+
+<dt>
+
+connectionName
+
+</dt>
+
+Reference to the database connection defined within the `config/database.ts` file. If not defined, we will use the default database connection.
+
+</dd>
+
+<dt>
+
+tableName
+
+</dt>
+
+<dd>
+
+The database table to use to store rate limits. 
+
+</dd>
+
+</dl>
+
+### Memory store
+
+The `memory` store is a simple in-memory store that can be useful for testing purposes but not only. Sometimes, for some use cases, you might want to have a lock that is only valid for the current process and not shared across multiple ones.
+
+The memory store is built on top of the [`async-mutex`](https://www.npmjs.com/package/async-mutex) package.
+
+```ts
+{
+  memory: stores.memory(),
+}
+```
+
+## Locking a resource
+
+Once you have configured your lock store, you can start using locks to protect your resources anywhere within your application.
+
+Here is a simple example of how to use locks to protect a resource.
+
+
+:::codegroup
+
+```ts
+// title: Manual locking
+import { errors } from '@adonisjs/lock'
+import locks from '@adonisjs/lock/services/main'
+import { HttpContext } from '@adonisjs/core/http'
+
+export default class OrderController {
+  public async process({ response, request }: HttpContext) {
+    const orderId = request.input('order_id')
+
+    const lock = locks.createLock(`order.processing.${orderId}`)
+
+    /**
+     * Try to acquire the lock immediately. Otherwise, throw an exception
+     * if the lock is already acquired.
+     */
+    try {
+      await lock.acquireImmediately()
+
+      /**
+       * Lock has been acquired. We can process the order
+       */
+      await processOrder()
+
+      /**
+       * Release the lock
+       */
+      await lock.release()
+      return 'Order processed successfully'
+    } catch (err) {
+      /**
+       * The lock is already locked by another process/request
+       */
+      if (err instanceof errors.E_LOCK_ALREADY_ACQUIRED) {
+        return 'Order is already being processed'
+      }
+    }
+  }
+}
+```
+
+```ts
+// title: Automatic locking
+import { errors } from '@adonisjs/lock'
+import locks from '@adonisjs/lock/services/main'
+import { HttpContext } from '@adonisjs/core/http'
+
+export default class OrderController {
+  public async process({ response, request }: HttpContext) {
+    const orderId = request.input('order_id')
+
+    /**
+     * Will run the function only if lock is available
+     * Lock will also be automatically released once the function
+     * has been executed
+     */
+    await locks
+      .createLock(`order.processing.${orderId}`)
+      .runImmediately(async (lock) => {
+        /**
+         * Lock has been acquired. We can process the order
+         */
+        await processOrder()
```

Comment:
Same here, what happens if `processOrder` throws exception?

Reply:
yup, if `processOrder` throws an exception, lock will be released here since we use `lock.run(() => {})` which automatically handles lock release.